execution_engine/src/storage/transaction_source/lmdb.rs

```diff
@@ -116,6 +155,39 @@ impl LmdbEnvironment {
     pub fn sync(&self) -> Result<(), lmdb::Error> {
         self.env.sync(true)
     }
+
+    /// Get LMDB environment info.
+    pub fn info(&self) -> Result<EnvInfo, lmdb::Error> {
+        let env: *mut MDB_env = self.env().env();
+        let mut env_info = mem::MaybeUninit::uninit();
+        unsafe {
+            lmdb_result(lmdb_sys::mdb_env_info(env, env_info.as_mut_ptr()))?;
+            let env_info = EnvInfo(env_info.assume_init());
+            Ok(env_info)
+        }
+    }
+
+    fn set_map_size(&self, map_size: usize) -> Result<(), lmdb::Error> {
+        let env: *mut MDB_env = self.env().env();
+        let ret = unsafe { lmdb_sys::mdb_env_set_mapsize(env, map_size) };
+        lmdb_result(ret)
+    }
+
+    fn resize_if_required(&self) -> Result<(), lmdb::Error> {
+        let env_info = self.info()?;
+        let stat = self.env.stat()?;
+        let size_used_bytes = stat.page_size() as usize * env_info.last_pgno();
+        let size_left_bytes = env_info.mapsize() - size_used_bytes;
+        if size_left_bytes <= self.grow_size_threshold {
+            let new_map_size = size_used_bytes + self.grow_size_bytes;
+            self.set_map_size(new_map_size)?;
+            debug!(
```

Comment:
I think `info` here is fine, it should be happening very rarely but when it does happen it would be nice to know it. We could also measure the time it took and log it. Again, it should be happening very rarely so won't pollute the logs.